src/proxy/buffer.rs

```diff
@@ -95,72 +147,383 @@ where
+    F: Clone,
+{
+    fn deadline(&self, req: &Req) -> Option<Instant> {
+        (self)(req)
+    }
+}
+
+impl<Req> Deadline<Req> for Duration {
+    fn deadline(&self, _: &Req) -> Option<Instant> {
+        Some(clock::now() + *self)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::future;
+    use futures::sync::oneshot::{self, Receiver, Sender};
+    use svc::Service;
+
+    struct Idle(Arc<()>);
+    impl svc::Service<()> for Idle {
+        type Response = ();
+        type Error = Error;
+        type Future = future::Empty<(), Error>;
+
+        fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+            Ok(Async::NotReady)
+        }
+
+        fn call(&mut self, _: ()) -> Self::Future {
+            unimplemented!("what you do?!")
+        }
+    }
+
+    struct Active(Option<Sender<Sender<()>>>);
+    impl svc::Service<()> for Active {
+        type Response = ();
+        type Error = Error;
+        type Future = Rsp;
+
+        fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+            Ok(Async::Ready(()))
+        }
+
+        fn call(&mut self, _: ()) -> Self::Future {
+            let (tx, rx) = oneshot::channel();
+            self.0
+                .take()
+                .expect("lost sender")
+                .send(tx)
+                .expect("rx must be alive");
+            Rsp(rx)
+        }
+    }
+
+    struct Rsp(Receiver<()>);
+    impl Future for Rsp {
+        type Item = ();
+        type Error = Error;
+        fn poll(&mut self) -> Poll<(), Error> {
+            Ok(self.0.poll().expect("must not fail"))
+        }
+    }
+
+    #[test]
+    fn request_aborted_with_idle_service() {
+        tokio::run(future::lazy(|| {
+            let mut svc = Enqueue::new(
+                Idle(Arc::new(())),
+                Duration::from_millis(100),
+                1,
+                &mut logging::context_executor("test"),
+            );
+
+            assert!(svc.poll_ready().ok().map(|r| r.is_ready()).unwrap_or(false));
+
+            svc.call(()).then(|r| match r {
+                Ok(_) => panic!("unexpected response from idle service"),
+                Err(e) => {
+                    assert!(e.downcast::<Aborted>().is_ok());
+                    future::ok(())
+                }
+            })
+        }));
+    }
+
+    #[test]
+    fn inner_service_dropped() {
+        tokio::run(future::lazy(|| {
+            let anchor = Arc::new(());
+            let handle = Arc::downgrade(&anchor);
+            let inner = Idle(anchor);
+            let mut svc = Enqueue::new(
+                inner,
+                Duration::from_secs(0),
+                1,
+                &mut logging::context_executor("test"),
+            );
+
+            assert!(svc.poll_ready().ok().map(|r| r.is_ready()).unwrap_or(false));
+            let call = svc.call(());
+            drop(svc);
+
+            let call = tokio::timer::Timeout::new(call, Duration::from_millis(100))
```

Comment:
TIOLI: Now that the [`tokio-test` crate](https://github.com/tokio-rs/tokio/tree/master/tokio-test) has been released, we might want to look into rewriting this (and other tests) to use mock timers instead. Maybe a follow-up?

Reply:
let's get an issue in the backlog for this?

Reply:
sounds good, i opened linkerd/linkerd2#2802